File: data-structures/queue/implementations/queue.ts
import { Queue } from "../interfaces/queue";


export class MyQueue<T> implements Queue<T> {
  private readonly initialCapacity: number;
  private currentCapacity: number;
  private container: Array<T>
  private front: number;
  private back: number;
  private currentLength: number


  public constructor(initialCapacity = 4) {
    if (initialCapacity <= 0) {
      throw new Error("initialCapacity must be greater than zero");
    }

    this.initialCapacity = initialCapacity;
    this.currentCapacity = initialCapacity;
    this.container = new Array(initialCapacity)
    this.front = 0
    this.back = 0
    this.currentLength = 0;
  }

  public length(): number {
    return this.currentLength;
  }

  public capacity(): number {
    return this.currentCapacity
  }

  public isEmpty(): boolean {
    return this.length() <= 0;
  }

  public peek(): T | undefined {
   return this.container[this.front]
  }

  public enqueue(_value: T): void {
    this.ensureCapacity();
    this.container[this.back] = _value

    this.back++
    this.currentLength++
  }

  public dequeue(): T | undefined {
    if(this.isEmpty()) return undefined;
    const node = this.container[this.front];

    for(let i = 1; i < this.length(); i++) {
      this.container[i-1] = this.container[i];
    }
    delete this.container[this.currentLength-1];
    this.currentLength--
    this.back--
    return node
  }

  public clear(): void {
    this.container = new Array(this.initialCapacity);
    this.currentLength = 0
  }

  public toArray(): T[] {
    const result: T[] = []

    for(const v of this.container) {
      if(v && v !== undefined) {
        result.push(v)
      }
    }

    return result
  }


  private isFull() {
    return this.capacity() === this.length()
  }

  private ensureCapacity() {

    if(!this.isFull()) {
      return
    }

    const newCapacity = this.length() * 2;
    const newContainer = new Array(newCapacity);


    for(let i = 0; i < this.length(); i++) {
      newContainer[i] = this.container[i]
    }

    this.container = newContainer
    this.currentCapacity = newCapacity
  }

  protected getInitialCapacity(): number {
    return this.initialCapacity;
  }
}

const queue = new MyQueue<number>(2)
queue.enqueue(1);
queue.enqueue(2);
queue.dequeue();
queue.enqueue(3);
queue.enqueue(4);
queue.dequeue();
queue.enqueue(5);



console.log("queue", queue);
// console.log(queue.length() === 4)

File: data-structures/queue/implementations/queue.go
package implementations

import (
	"dsa/data-structures/queue/interfaces"
)

type MyQueue[T any] struct {
	initialCapacity int
	currentCapacity int
	currentLength   int
	head            int
	tail            int
	container       []T
}

var _ interfaces.Queue[int] = (*MyQueue[int])(nil)

func NewMyQueue[T any](initialCapacity int) *MyQueue[T] {
	if initialCapacity <= 0 {
		panic("initialCapacity must be greater than zero")
	}

	return &MyQueue[T]{
		initialCapacity: initialCapacity,
		container:       make([]T, initialCapacity),
		currentCapacity: initialCapacity,
		currentLength:   0,
		head:            0,
		tail:            0,
	}
}

func (q *MyQueue[T]) Length() int {
	return q.currentLength
}

func (q *MyQueue[T]) Capacity() int {
	return q.currentCapacity
}

func (q *MyQueue[T]) IsEmpty() bool {
	return q.Length() <= 0
}

func (q *MyQueue[T]) Peek() T {
	if q.IsEmpty() {
		panic("peek: queue is empty")
	}
	return q.container[q.head]
}

func (q *MyQueue[T]) Enqueue(value T) {
	q.ensureCapacity()
	q.container[q.tail] = value

	q.tail++
	q.currentLength++
}

func (q *MyQueue[T]) Dequeue() T {
	if q.IsEmpty() {
		panic("dequeue: queue is empty")
	}
	val := q.container[q.head]

	q.shift() // O(N)
	q.currentLength--
	q.tail--
	return val
}

func (q *MyQueue[T]) Clear() {
	q.container = nil
	q.currentLength = 0
	q.currentCapacity = 0
	q.head = 0
	q.tail = 0
}

func (q *MyQueue[T]) ToSlice() []T {
	rs := make([]T, 0, q.Length())

	for i := 0; i < q.Length(); i++ {
		rs = append(rs, q.container[i])
	}

	return rs
}

func (q *MyQueue[T]) InitialCapacity() int {
	return q.initialCapacity
}

func (q *MyQueue[T]) IsFull() bool {
	return q.Length() == q.Capacity()
}

func (q *MyQueue[T]) ensureCapacity() {
	if !q.IsFull() {
		return
	}

	newCap := q.currentCapacity * 2
	newCon := make([]T, newCap)

	for i := 0; i < q.Length(); i++ {
		newCon[i] = q.container[i]
	}
	// copy(newCon, q.container)

	q.currentCapacity = newCap
	q.container = newCon
}

func (q *MyQueue[T]) shift() {
	for i := 1; i < q.Length(); i++ {
		q.container[i-1] = q.container[i]
	}
}
